### PR TITLE
Add noupdateserviceaccount policy to library

### PR DESCRIPTION
**Pull request by mtaufen**
This policy helps prevent workloads with "update-self" permissions from
escalating further in the cluster by selecting a new service account to
run as. It is especially useful for workloads running in sensitive
namespaces like kube-system, where nearby service accounts are likely to
have cluster admin rights.

@maxsmythe

### CODE REVIEW

File: library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml
Comment (mtaufen) [inline library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: I'm not really sure how these tests work. Do we have an emulated kubernetes API such that we can test mutations of existing resources, or is it a stateless test?

Comment (maxsmythe) [reply library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: Unfortunately the tests are essentially stateless (only CREATE is tested). For our local tests, we are looking at ways to provide a full K8s AdmissionReview object so that things like "only-on-update" can be appropriately tested.

Generally edge-based enforcement is discouraged if there is an alternative (only works at request time, does not work for audit), but because this is specifically testing changes to the service account it seems reasonable.

Comment (mtaufen) [reply library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: Got it - what's the best thing to do here? Just comment out the "disallowed" example from `suite.yaml`?

For level-triggered audits like this, you _might_ be able to extract past revisions of Deployments, etc. They have to be stored _somewhere_ since there's a rollback option: https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#checking-rollout-history-of-a-deployment

Comment (maxsmythe) [reply library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: Yeah, commenting out for now seems like the best course.

re: level-triggered audits... interesting. I wonder where this data lives? Currently we are scraping the objects as they exist on the cluster.

**Review by maxsmythe — COMMENTED** [review]
Thanks for the PR!

Code LGTM, but we'll also need the Rego source in the src/ directory, like so:

https://github.com/open-policy-agent/gatekeeper-library/tree/master/src/general/allowedrepos

It'll also be possible to cover the edge-based unit tests using rego tests.

**Comment by ritazh** [comment]
@mtaufen Thanks for the PR! 
In addition to Rego source in the src/ directory, a recent automation change requires the creation of 
> A constraint template named `src/<policy-name>/constraint.tmpl` with a `description` annotation and the parameter structure.

e.g. https://github.com/open-policy-agent/gatekeeper-library/blob/master/src/general/allowedrepos/constraint.tmpl

PTAL https://github.com/open-policy-agent/gatekeeper-library#new-policy for more details.

@chewong should `make generate` fail if `constraint.tmpl` is missing for a policy in `src/`?

**Comment by mtaufen** [comment]
@ritazh @chewong FYI. I ran `make generate` but it actually deleted the contents of a bunch of template files:

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   library/general/allowedrepos/template.yaml
        modified:   library/general/automount-serviceaccount-token/template.yaml
        modified:   library/general/block-endpoint-edit-default-role/template.yaml
        modified:   library/general/block-nodeport-services/template.yaml
        modified:   library/general/containerlimits/template.yaml
        modified:   library/general/containerresourceratios/template.yaml
        modified:   library/general/disallowanonymous/template.yaml
        modified:   library/general/disallowedtags/template.yaml
        modified:   library/general/externalip/template.yaml
        modified:   library/general/httpsonly/template.yaml
        modified:   library/general/imagedigests/template.yaml
        modified:   library/general/noupdateserviceaccount/template.yaml
        modified:   library/general/replicalimits/template.yaml
        modified:   library/general/requiredannotations/template.yaml
        modified:   library/general/requiredlabels/template.yaml
        modified:   library/general/requiredprobes/template.yaml
        modified:   library/general/uniqueingresshost/template.yaml
        modified:   library/general/uniqueserviceselector/template.yaml
        modified:   library/pod-security-policy/allow-privilege-escalation/template.yaml
        modified:   library/pod-security-policy/apparmor/template.yaml
        modified:   library/pod-security-policy/capabilities/template.yaml
        modified:   library/pod-security-policy/flexvolume-drivers/template.yaml
        modified:   library/pod-security-policy/forbidden-sysctls/template.yaml
        modified:   library/pod-security-policy/fsgroup/template.yaml
        modified:   library/pod-security-policy/host-filesystem/template.yaml
        modified:   library/pod-security-policy/host-namespaces/template.yaml
        modified:   library/pod-security-policy/host-network-ports/template.yaml
        modified:   library/pod-security-policy/privileged-containers/template.yaml
        modified:   library/pod-security-policy/proc-mount/template.yaml
        modified:   library/pod-security-policy/read-only-root-filesystem/template.yaml
        modified:   library/pod-security-policy/seccomp/template.yaml
        modified:   library/pod-security-policy/selinux/template.yaml
        modified:   library/pod-security-policy/users/template.yaml
        modified:   library/pod-security-policy/volumes/template.yaml
```

For example:

```
diff --git a/library/general/allowedrepos/template.yaml b/library/general/allowedrepos/template.yaml
index 23c0c7d..e69de29 100644
--- a/library/general/allowedrepos/template.yaml
+++ b/library/general/allowedrepos/template.yaml
@@ -1,40 +0,0 @@
-apiVersion: templates.gatekeeper.sh/v1beta1
-kind: ConstraintTemplate
-metadata:
-  name: k8sallowedrepos
-  annotations:
-    description: >-
-      Requires container images to begin with a string from the specified list.
-spec:
-  crd:
-    spec:
-      names:
-        kind: K8sAllowedRepos
-      validation:
-        # Schema for the `parameters` field
-        openAPIV3Schema:
-          type: object
-          properties:
-            repos:
-              description: The list of prefixes a container image is allowed to have.
-              type: array
-              items:
-                type: string
-  targets:
-    - target: admission.k8s.gatekeeper.sh
-      rego: |
-        package k8sallowedrepos
-
-        violation[{"msg": msg}] {
-          container := input.review.object.spec.containers[_]
-          satisfied := [good | repo = input.parameters.repos[_] ; good = startswith(container.image, repo)]
-          not any(satisfied)
-          msg := sprintf("container <%v> has an invalid image repo <%v>, allowed repos are %v", [container.name, container.image, input.parameters.repos])
-        }
-
-        violation[{"msg": msg}] {
-          container := input.review.object.spec.initContainers[_]
-          satisfied := [good | repo = input.parameters.repos[_] ; good = startswith(container.image, repo)]
-          not any(satisfied)
-          msg := sprintf("initContainer <%v> has an invalid image repo <%v>, allowed repos are %v", [container.name, container.image, input.parameters.repos])
-        }
```

**Comment by chewong** [comment]
Did you install [`gomplate`](https://docs.gomplate.ca/installing/) to your PATH?

**Comment by mtaufen** [comment]
> Did you install [gomplate](https://docs.gomplate.ca/installing/) to your PATH?

Thanks, that was it!

I added tests and pushed some updates PTAL @ritazh @maxsmythe @chewong 

Also I ended up using `strings.Indent 6` instead of `strings.Indent 8` (from the [example](https://github.com/open-policy-agent/gatekeeper-library/blob/master/src/general/allowedrepos/constraint.tmpl) you gave) in my `constraint.tmpl` since that matched the original yaml indentation I had. LMK if you prefer `strings.Indent 8`, I just didn't want to anger the yaml indentation gods by changing more than necessary.

File: library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml
Comment (mtaufen) [reply library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: Ack, done.

File: src/general/noupdateserviceaccount/src.rego
Comment (maxsmythe) [inline src/general/noupdateserviceaccount/src.rego:74]: nit: newline

File: src/general/noupdateserviceaccount/src_test.rego
Comment (maxsmythe) [inline src/general/noupdateserviceaccount/src_test.rego:318]: nit: `true` superfluous here and above

**Review by maxsmythe — APPROVED** [review]
LGTM! 2 small nits

File: src/general/noupdateserviceaccount/src.rego
Comment (mtaufen) [reply src/general/noupdateserviceaccount/src.rego]: Done/

File: src/general/noupdateserviceaccount/src_test.rego
Comment (mtaufen) [reply src/general/noupdateserviceaccount/src_test.rego]: Done.

File: library/general/noupdateserviceaccount/template.yaml
Comment (ritazh) [inline library/general/noupdateserviceaccount/template.yaml:32]: Since this rego looks for `username` from `userInfo`, it will definitely fail when in audit, but since it also only applies to `UPDATE` operations, which is also not available in audit, it's essentially a pass thru policy. We should document this policy does NOT work for audit.

File: library/general/noupdateserviceaccount/suite.yaml
Comment (ritazh) [inline library/general/noupdateserviceaccount/suite.yaml:15]: @willbeason Would this type of policy that requires `input.review.oldObject`  need to be tested with `gator test`?

Comment (maxsmythe) [reply library/general/noupdateserviceaccount/suite.yaml:15]: I think yes

Comment (willbeason) [reply library/general/noupdateserviceaccount/suite.yaml:15]: Yeah - we're considering allowing test authors to optionally specify a full AdmissionReview instead of an object so that they can specify all of the other fields

File: library/general/noupdateserviceaccount/template.yaml
Comment (willbeason) [reply library/general/noupdateserviceaccount/template.yaml:32]: That's a really good point - it gives me a question unrelated to the PR so I'll ask in gatekeeper-maintainers.

Comment (mtaufen) [reply library/general/noupdateserviceaccount/template.yaml:32]: Where is the best place to document that? Just in the docstrings?

File: library/general/noupdateserviceaccount/suite.yaml
Comment (mtaufen) [reply library/general/noupdateserviceaccount/suite.yaml:15]: Do I need to add something to the PR for that?

Comment (willbeason) [reply library/general/noupdateserviceaccount/suite.yaml:15]: Nope - this is a side design discussion for something completely different. No action necessary on your part!

Comment (willbeason) [reply library/general/noupdateserviceaccount/suite.yaml:15]: Filed https://github.com/open-policy-agent/gatekeeper/issues/1843

**Review by willbeason — APPROVED** [review]

File: library/general/noupdateserviceaccount/template.yaml
Comment (sozercan) [reply library/general/noupdateserviceaccount/template.yaml:32]: we can add in annotation and readme in the `noupdateserviceaccount` folder perhaps

Comment (mtaufen) [reply library/general/noupdateserviceaccount/template.yaml:32]: Thanks, added.

**Review by ritazh — APPROVED** [review]
lgtm

**Comment by mtaufen** [comment]
Fixed that one, but now it wants the `example_*` files to each contain exactly one object. Should I just stick placeholder Deployments in each or something, since we can't really provide an example with only one object for this specific policy?

```
    --- FAIL: example-allowed	(0.066s)
        object file must contain exactly one object: "usr/local/google/home/mtaufen/proj/gatekeeper-library/library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml" defines 4 objects

```

**Comment by maxsmythe** [comment]
Yeah, placeholder sounds fine.

File: library/general/noupdateserviceaccount/samples/noupdateserviceaccount/constraint.yaml
Comment (ritazh) [inline library/general/noupdateserviceaccount/samples/noupdateserviceaccount/constraint.yaml:32]: 👍

**Comment by mtaufen** [comment]
Any ideas why https://github.com/open-policy-agent/gatekeeper-library/runs/5146018047?check_suite_focus=true failed?

I saw this in the logs but looks unrelated to this PR:
```
# error: unable to recognize "library/general/automount-serviceaccount-token/samples/automount-serviceaccount-token/constraint.yaml": no matches for kind "K8sPSPAutomountServiceAccountTokenPod" in version "constraints.gatekeeper.sh/v1beta1"
```

**Comment by maxsmythe** [comment]
That looks like for some reason a CRT didn't get created for the automount service account token CT. That could be due to timeouts or a bug in the CT (the joys of testing against an eventually consistent system).

File: library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml
Comment (willbeason) [reply library/general/noupdateserviceaccount/samples/noupdateserviceaccount/example_allowed.yaml:1]: Yep - tracking in https://github.com/open-policy-agent/gatekeeper/issues/1843

**Comment by ritazh** [comment]
```
# error: unable to recognize "library/general/automount-serviceaccount-token/samples/automount-
```
we can ignore retry errors like these

This is not a flake. seems the `noupdateserviceaccount` policy is failing the integration test with:
```
# deployment.apps "policy-test" deleted
# expected: denied the request
# actual: deployment.apps/policy-test created
```
which is caused by https://github.com/open-policy-agent/gatekeeper-library/blob/master/test/bats/test.bats#L102 since the integration test will look for `example-disallowed.yaml` and expect it to be blocked. But the rego will let it thru because it's not an update since the `example-allowed.yaml` policy-test resources have been [deleted](https://github.com/open-policy-agent/gatekeeper-library/blob/master/test/bats/test.bats#L86). I don't think the current integration test works for any referential constraints.

To unblock this PR, we could update test.bats to conditionally skip/delay deletion of allowed resources for certain policies OR remove the disallowed yaml for this policy. Do folks have any preference?

**Comment by maxsmythe** [comment]
IMO let's remove disallowed, since the resource itself is not technically disallowed, only in the context of updating the allowed resource. Because we don't currently have a way to broadcast that, it makes the example misleading.

**Comment by maxsmythe** [comment]
re: error line... oops for over-fixating on that, I wonder if we should add a "retrying due to error" context so people know it's a handled error.

**Comment by mtaufen** [comment]
Sweet, deleted the file. Let's try again :)

**Pull request merged**
